### PR TITLE
Remove track tab and rounded corners

### PR DESCRIPTION
**Pull request by pierricgimmig**
This PR's goal is to gain vertical real-estate by removing track tabs in
favor of a track header that is now aligned with tracks, to their left.
This is also similar to how most track UI applications display tracks. 

Note that there are many things we could do to enhance the look of 
tracks, like fine tuning text and collapsible arrow position, coloring on
hover, better separation of header and track. All those changes are 
welcome but outside the scope of this PR. This basically takes our existing
tabs and moves them to the left of the tracks.

image

### CODE REVIEW

File: src/OrbitGl/CaptureWindow.cpp
Comment (beckerhe) [inline src/OrbitGl/CaptureWindow.cpp:734]: I guess the old calls can now be removed?

**Review by beckerhe — APPROVED** [review]
Awesome. This is a cool change. I like the new layout. Code-wise this also looks good to me, but be aware that I don't know a lot about the TimeGraph code. It would be great if you could wait for a review from Florian or Pablo as well. I added one comment, otherwise LGTM.

One more thing: The labels in the timers in your screenshot seem to start before the actual timer box. Is that an unrelated bug or could this be triggered by your margin changes?

File: src/OrbitGl/Track.cpp
Comment (florian-kuebler) [inline src/OrbitGl/Track.cpp:91]: Again, I am a bit confused that the layout's margin is used for padding inside the track.

File: src/OrbitGl/CaptureWindow.cpp
Comment (florian-kuebler) [reply src/OrbitGl/CaptureWindow.cpp:734]: well those are still used here, right?

Comment (beckerhe) [reply src/OrbitGl/CaptureWindow.cpp:734]: Ohhh, I get it now. Sorry never mind.

You can consider writing it in a single expression (or not - what you deem is best).

**Review by reichlfl — COMMENTED** [review]
Hey Pierric, thanks for the change! Looks good, I have only minor comments. But before I approve, could you please check two things:

1) I can't judge how this looks for sub-tracks, could you provide a screenshot?
2) Does the accessibility for all elements still work correctly? Unfortunately this can only be tested on Windows afaik (using e.g. the Accessibility Insights tool and verifying that all the bounding boxes of the tracks and headers are correct).

As for 2, if this is not the case (or problematic to test) we need to discuss if we still need this as it has only been used for E2E test automation.

File: src/OrbitGl/TimeGraph.cpp
Comment (pierricgimmig) [reply src/OrbitGl/TimeGraph.cpp:906]: This is for the small portion at the left of the timeline and slider. It is possible for tracks to draw in those regions when scrolling vertically. Previously that was hidden by a single rectangle; now we need two because the tracks don't need to have a second masking rectangle to their left, they already have the header, which also leaves space between tracks unmasked, which is nice to have.

File: src/OrbitGl/TimeGraph.cpp
Comment (pierricgimmig) [reply src/OrbitGl/TimeGraph.cpp:911]: Not the slider, but track content can spill to that area when scrolling vertically, see comment above.

File: src/OrbitGl/TrackHeader.cpp
Comment (pierricgimmig) [reply src/OrbitGl/TrackHeader.cpp:94]: will look into it, yep, thanks.

File: src/OrbitGl/include/OrbitGl/TrackContainer.h
Comment (pierricgimmig) [reply src/OrbitGl/include/OrbitGl/TrackContainer.h:116]: it's not, removed, thanks

**Comment by pierricgimmig** [comment]
> Does the accessibility for all elements still work correctly? 

@reichlfl , it looks like it:

header_access
track_access
access_test

File: src/OrbitGl/Track.cpp
Comment (pierricgimmig) [reply src/OrbitGl/Track.cpp:91]: changed to TrackHeaderWidth()

File: src/OrbitGl/TimeGraph.cpp
Comment (pierricgimmig) [reply src/OrbitGl/TimeGraph.cpp:738]: changed to TrackHeaderWidth()

**Comment by reichlfl** [comment]
> > Does the accessibility for all elements still work correctly?
> 
> @reichlfl , it looks like it:

Cool, thanks for checking! Can you check the `TimerPane` as well? That should be the area which only contains the timers, it might now span the whole track. That would probably be an easy fix within the `AccessibleTrack` implementation - if not, I also think it's not a big deal because the tests (AFAIK) only test for existence, not exact extents of this.

**Comment by pierricgimmig** [comment]
> I can't judge how this looks for sub-tracks, could you provide a screenshot?

@reichlfl , @florian-kuebler , how can I test this? Memory track?

**Comment by pierricgimmig** [comment]
> One more thing: The labels in the timers in your screenshot seem to start before the actual timer box. Is that an unrelated bug or could this be triggered by your margin changes?

@beckerhe , I noticed that too, but this is unrelated to this change. I think this comes from the Qt text rendering change, I've noticed other artifacts like text overlapping sometimes.

File: src/OrbitGl/include/OrbitGl/TrackHeader.h
Comment (pierricgimmig) [reply src/OrbitGl/include/OrbitGl/TrackHeader.h:39]: Seems to work, see screenshots below.

**Comment by reichlfl** [comment]
> 

Memory track is probably the easiest, yes

**Comment by pierricgimmig** [comment]
> 

@reichlfl , the memory track seems a bit broken, thanks for pointing this out, will fix soon. Cheers!

**Comment by pierricgimmig** [comment]
The memory track issue was the legend being drawn outside the track, it is fixed now:

file:///home/pierric/Pictures/Screenshots/mem_track.png

I still have to check for track that contain children, not sure how to do that easily, I'll try setting up a hierarchy in code.

**Comment by pierricgimmig** [comment]
@reichlfl , @florian-kuebler, I don't see how I can test subtracks on a generic Linux capture, would you know how? Would you have some test data perhaps, or maybe you could compile this branch and test on your side? Thanks.

**Comment by florian-kuebler** [comment]
https://github.com/google/orbit/blob/main/contrib/automation_tests/testdata/OrbitTest_1-72.orbit could be a good example.

But, Ill also check the gpu tracks tomorrow.

**Review by florian-kuebler — CHANGES_REQUESTED** [review]
Actually there are some issues with subtracks left.

1. On the OrbitTest_1-72 capture the page fault track shows some issues. Subtrack heads is on the tracks header and also the legend is drawn on top of the data. See screenshot:
Screenshot 2022-12-15 at 16 25 03

2. There is a crash for captures with the gpu (subtrack). I will look into this, briefly, but fear that I don't have a shareable capture at hand. (I will try to get something though).
Screenshot 2022-12-15 at 16 33 06

In general, I like that we are trying to get more vertical space. But also I feel that the space on the left now is pretty big, considering we still have the split between timeline and the other data views.
Couldn't we do the middle ground: Have the bare minimum needed for the track headers. Like one or two pixels between tracks; remove the rounded corners for the header and also make the headers smaller?

**Comment by pierricgimmig** [comment]
Thanks for checking @reichlfl . I'll try out the data and fix the issues. Concerning the big headers, I also found the same at first, but got really accustomed to it using other tools that have this layout. I think it might be a great place to put extra information about track stats too. One thing I want to add is the ability to resize the header part, so we could go from what is suggested to what you are suggesting when minimizing the header width.

**Comment by florian-kuebler** [comment]
>One thing I want to add is the ability to resize the header part, so we could go from what is suggested to what you are suggesting when minimizing the header width.

But then why scraping out the current header, when we are going to add it back again? Maybe we should collect a bunch of screenshots from other tools with this track like format, and see if we find a common sense (as said, I am a big fan of reducing vertical gaps). I looked at tools like final cut pro (which has the header) and fruity loops (which has the left part, but much smaller).

File: src/OrbitGl/include/OrbitGl/TrackHeader.h
Comment (florian-kuebler) [inline src/OrbitGl/include/OrbitGl/TrackHeader.h:39]: That one here is actually the issue that leads to the crash.

Basically `GpuSubmissionTrack::GetHeight` is calling `header_->GetHeight()`, which again with this line here is calling `GpuSubmissionTrack::GetHeight` (as the track is the parent of the header.

Comment (florian-kuebler) [reply src/OrbitGl/include/OrbitGl/TrackHeader.h:39]: @reichlfl FYI

Comment (florian-kuebler) [reply src/OrbitGl/include/OrbitGl/TrackHeader.h:39]: Adding a quick fix (which should not be the way to go), we see the same issues as in the memory track:
Screenshot 2022-12-16 at 11 58 03

**Comment by florian-kuebler** [comment]
Just for reference. This is how a more condensed ui with the current implementation logic could look like:
Screenshot 2022-12-16 at 12 51 33

I made all the gaps smaller, removed the rounding, reduced the size of the triangle and shifted the header to the left (including reducing the width).

**Comment by florian-kuebler** [comment]
Also, I extracted the property controls from here to https://github.com/google/orbit/pull/4600, as it is nice for testing/playing around.

**Comment by pierricgimmig** [comment]
@florian-kuebler , thanks looking into this, I'll fix the remaining issues with the left headers. And then we can compare the different  approaches.

**Comment by pierricgimmig** [comment]
@florian-kuebler , I thought about this a bit more, I think having the track headers to the left is really the way to go. It makes everything clearer, there is a room for additional info and icons (a pin for example, or a visibility toggle), and it's just a cleaner layout. I don't think horizontal real estate is as important as vertical real estate. I thought it was when I originally implemented the track tabs, but it's not. I think Perfetto is a really good example at a clean layout that we should move towards. There are many other examples of track UI that use the same layout. I'll continue fixing the remaining bugs in the current PR. Please let me know if any internal discussion oppose the current change. Thanks.

**Comment by pierricgimmig** [comment]
@reichlfl , @florian-kuebler , the PR is in a better state, I'd love to get your feedback. Thanks!

**Comment by florian-kuebler** [comment]
> @reichlfl , @florian-kuebler , the PR is in a better state, I'd love to get your feedback. Thanks!

Thanks. I'll have a look and test it tomorrow.

File: src/OrbitGl/include/OrbitGl/TrackHeader.h
Comment (florian-kuebler) [reply src/OrbitGl/include/OrbitGl/TrackHeader.h:39]: @pierricgimmig This crash here does not seem to be fixed.

**Comment by pierricgimmig** [comment]
Thanks for the capture @ronaldfw!

File: src/OrbitGl/include/OrbitGl/TrackHeader.h
Comment (pierricgimmig) [reply src/OrbitGl/include/OrbitGl/TrackHeader.h:39]: Ah ok, hopefully I can repro with @ronaldfw 's capture.

Comment (florian-kuebler) [reply src/OrbitGl/include/OrbitGl/TrackHeader.h:39]: So, I looked into your fix for the crash. I still think this is against the layouting principles we came up with. The idea about our parent/child relationship is, that the position and size is controlled by the parent (this is mostly due to the lack of flexible layouting), and we tried to be consistent about this.

Now with your change you must remember that you can't call `GetHeight` for you the header, while it is fine and maybe even required to call the same method for your other childs (like for subtracks).

Maybe @reichlfl has a better idea how to implement the `TrackHeader::GetHeight` method.

**Comment by florian-kuebler** [comment]
Thanks for the changes so far. I played around with the current state, and mostly like it.

Despite my comment on the `TrackHeader::GetHeight` method, which I think must be resolved, I found three remaining bugs:
1. If you expand the GpuTrack and then collapse the submission subtrack, the header position is off (see screenshot). The same happens, if the submissions only have a depth of 1.
Screenshot 2022-12-21 at 08 13 13
1. Similarly, if you have a track with gpu debug markers and have that one collapsed, the position of the header is also off:
 Screenshot 2022-12-21 at 08 21 30
1. If you have empty thread tracks, that only have thread state, there is a spacing at the bottom missing:
Screenshot 2022-12-21 at 08 17 57

Also, I do like the new indicators for expand/collapse. I am wondering whether they should be swapped, however. @reichlfl what are your thoughts on that?

Finally, I would decrease the width of the header from 350 to 250.

File: src/OrbitGl/TrackHeader.cpp
Comment (florian-kuebler) [inline src/OrbitGl/TrackHeader.cpp:49]: Also here should probably not use the parent's height.

File: src/OrbitGl/GpuTrack.cpp
Comment (florian-kuebler) [inline src/OrbitGl/GpuTrack.cpp:98]: I guess these two changes here result in the problem we've seen with the subtrack's labels being off (or better with the tracks not aligned to the headers).

I guess we want to add a `header_->GetHeight()` here and below, which requires a different implementation of `TrackHeader::GetHeight`.

File: src/OrbitGl/include/OrbitGl/TrackHeader.h
Comment (florian-kuebler) [reply src/OrbitGl/include/OrbitGl/TrackHeader.h:39]: I gave this another thought. The track header is completely inside the track (which is correct). It's relevant height however, is not the complete track's height, but only the height of the label (or the toggle) + some slight spacing at the top and below the label.

So you can implement this method with information locally known here.

This also allows to fix the issues with the header being out of the track in case of subtracks, as we can move the subtracks below the parent's header when the parent track is expanded.

Comment (florian-kuebler) [reply src/OrbitGl/include/OrbitGl/TrackHeader.h:39]: I guess the easy fix would be to explicitly set the header's height in `Track::DoUpdateLayout()`.

Alternatively, @reichlfl mentioned that we could do something similar to the capture view element's `width` using the `LayoutFlags`.

But, I guess the simple solution would work for now.

**Comment by pierricgimmig** [comment]
Hi @florian-kuebler , thanks for your comments. I fixed 1, 2 and 3. I didn't want to spend too much time on 2 as I think the subtracks should be completely overhauled in favor of a full track hierarchy where subtracks are treated as full-fledged tracks, with their own header. Thanks again for your reviews.

File: src/OrbitGl/include/OrbitGl/TrackHeader.h
Comment (reichlfl) [reply src/OrbitGl/include/OrbitGl/TrackHeader.h:39]: @pierricgimmig I'm not sure if this discussion happened before or after you requested the last review - just to be sure this doesn't get lost, as far as I see it this is the main conceptual issue that prevents this from being merged. Children cannot use the parent height to determine their own height, only the other way around. Please instead introduce a `SetHeight` method for the headers, and call this during `Track::DoUpdateLayout`. Make sure to set the "update required" flags accordingly within `SetHeight` (see other layout-changing setters for reference).

**Comment by reichlfl** [comment]
> I am wondering whether they should be swapped, however

Apparently the current version is also what the latest Material UI accordion elements use. Even though I find this confusing, I guess we should stick with it. To be sure: "arrow down" means it's currently collapsed, "arrow up" means it's expanded, right?

File: src/OrbitGl/ThreadTrack.cpp
Comment (karupayun) [inline src/OrbitGl/ThreadTrack.cpp:275]: Since you are always using `pos[0] + header_->GetWidth()` and to avoid repetition of code, I would prefer to include the header_width in the variable pos.

File: src/OrbitGl/TimeGraph.cpp
Comment (karupayun) [inline src/OrbitGl/TimeGraph.cpp:885]: Why we need this?

File: src/OrbitGl/TrackHeader.cpp
Comment (karupayun) [inline src/OrbitGl/TrackHeader.cpp:57]: What do you think about choosing a slightly different color for the TrackTab to differentiate it with its content?

File: src/OrbitGl/TrackHeader.cpp
Comment (karupayun) [inline src/OrbitGl/TrackHeader.cpp:107]: It would be better to include this offset in the calculation of the position (`toggle_y_pos`)

File: src/OrbitGl/TrackHeader.cpp
Comment (karupayun) [inline src/OrbitGl/TrackHeader.cpp:97]: Could you explain why this offsets with a comment?

File: src/OrbitGl/TrackHeader.cpp
Comment (karupayun) [inline src/OrbitGl/TrackHeader.cpp:115]: This only works when IndentationLevel is 1. Please solve the issue or include that information in the comment and log an error in case that IndentationLevel is higher than 1.

File: src/OrbitQt/TimeGraphLayoutWidget.cpp
Comment (karupayun) [inline src/OrbitQt/TimeGraphLayoutWidget.cpp:35]: You should also erase the widget for track_tab_width_, track_tab_height_ and track_tab_offset_, or I am missing something?

**Review by karupayun — COMMENTED** [review]
Nice PR Pierric! At first I was unsure of the new layout, but I think that I'm ok with it now. I would include a different color for the header and the content, so it's easier to understand the border. What do you think?

A few more details:
1-
7YcUWgzarMKeBHX

The green line with the mouse can be seen a bit more to the left than the content, do you know why is that?

2-
5HuS5jW7fr4v3Ko

The toggles of the subtracks seems a bit broken.


3-
E6nDRVz8PPNoom9

The title of the FrameTrack intersect with the average time in the most common case, could you move this line to be only drawn in the content part?

4-
5FNmxP4saaDNM92

There are some broken cases with the Gpu-subtracks. Here it's an example with the track expanded but both subtracks contracted.

**Comment by pierricgimmig** [comment]
@reichlfl , @karupayun, thanks for the review and comments, I'll address them soon, I might revert the arrows for now too. As pointed out, they seem to render weirdly sometimes. I don't fully understand why, but it seems primitives are sometimes off by one pixel.

File: src/OrbitGl/include/OrbitGl/TrackHeader.h
Comment (pierricgimmig) [reply src/OrbitGl/include/OrbitGl/TrackHeader.h]: Implemented your suggestion regarding a SetHeight method @reichlfl

File: src/OrbitGl/TrackHeader.cpp
Comment (pierricgimmig) [reply src/OrbitGl/TrackHeader.cpp:49]: done.

File: src/OrbitGl/TimeGraph.cpp
Comment (pierricgimmig) [reply src/OrbitGl/TimeGraph.cpp:885]: For debugging

Comment (pierricgimmig) [reply src/OrbitGl/TimeGraph.cpp:885]: What I mean is that it's a useful debugging feature, the code is meant to be merged.

File: src/OrbitGl/TrackHeader.cpp
Comment (pierricgimmig) [reply src/OrbitGl/TrackHeader.cpp:97]: this was an attempt to fix the asymmetry, but it doesn't work all the time, so I'm removing it.

File: src/OrbitQt/TimeGraphLayoutWidget.cpp
Comment (pierricgimmig) [reply src/OrbitQt/TimeGraphLayoutWidget.cpp:35]: good point, done, thanks.

File: src/OrbitGl/TrackHeader.cpp
Comment (pierricgimmig) [reply src/OrbitGl/TrackHeader.cpp:107]: Done.

File: src/OrbitGl/ThreadTrack.cpp
Comment (pierricgimmig) [reply src/OrbitGl/ThreadTrack.cpp:275]: done.

File: src/OrbitGl/GraphTrack.cpp
Comment (pierricgimmig) [reply src/OrbitGl/GraphTrack.cpp:234]: done

File: src/OrbitGl/TrackHeader.cpp
Comment (pierricgimmig) [reply src/OrbitGl/TrackHeader.cpp:115]: Added a log as fixing this properly is a change that will come in its own PR, basically overhauling the track hierarchy system.

File: src/OrbitGl/TrackHeader.cpp
Comment (pierricgimmig) [reply src/OrbitGl/TrackHeader.cpp:57]: Yeah that would be cool, would have been easy to set colors with ImGui, but ImGui is not there anymore. Let's do this in an other PR.

File: src/OrbitGl/CaptureWindow.cpp
Comment (pierricgimmig) [reply src/OrbitGl/CaptureWindow.cpp]: simplified the code

**Comment by pierricgimmig** [comment]
@reichlfl , @karupayun, @florian-kuebler , I addressed most of the comments. Concerning the GPU subtracks issue, is that blocking? I feel like it might be faster if someone who knows the code better implemented a quick fix rather than going back and forth...  Also, I rebased this change in another PR if you want to see it with the latest UI changes. thanks for your help!

**Comment by pierricgimmig** [comment]
> The green line with the mouse can be seen a bit more to the left than the content, do you know why is that?

N/A in the rebase branch, and I couldn't repro on Windows weirdly.

**Comment by pierricgimmig** [comment]
> The toggles of the subtracks seems a bit broken.

Reverted, the ceiling code was the reason behind the weirdness.

**Comment by pierricgimmig** [comment]
> The title of the FrameTrack intersect with the average time in the most common case, could you move this line to be only drawn in the content part?

fixed

**Comment by florian-kuebler** [comment]
Thanks for addressing the comments. I think it looks good now, but I want to do another full review. Shall I do it here, or at the #4727 pr?

**Comment by pierricgimmig** [comment]
Closing as https://github.com/google/orbit/pull/4727 supersedes this PR.

**Pull request closed without merging**
